internal/message.go

```diff
@@ -0,0 +1,57 @@
+// Copyright The OWASP Coraza contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal
+
+import (
+	"fmt"
+	"net"
+)
+
+func getString(args map[string]interface{}, key string) (string, error) {
+	argVal, exist := args[key]
+	if exist && argVal != nil {
+		val, ok := argVal.(string)
+		if ok {
+			return val, nil
+		}
+	}
+
+	return "", fmt.Errorf("invalid argument for %s, string expected, got %v", key, argVal)
```

Comment:
IIUC, this error should be returned if not `ok` above, while here it is the not exists case, so not an error

Reply:
How to signal caller that there is no such key in the map/message? Is below code OK?
```

func getString(args map[string]interface{}, key string) (string, bool, error) {
	argVal, exist := args[key]
	if !exist {
		return "", false, nil
	}
	if argVal == nil {
		return "", true, nil
	}
	val, ok := argVal.(string)
	if !ok {
		return "", true, fmt.Errorf("invalid argument for %s, string expected, got %v", key, argVal)
	}
	return val, true, nil
}
```

Reply:
`coraza-spoa` requires `method` field in the message.
1. message `app=simple_app method=  path...` - variable value set up  issue in HAProxy (probably [haproxy.cfg](https://github.com/corazawaf/coraza-spoa/blob/main/docker/haproxy/haproxy.cfg);
2. message `app=simple_app path...` - message definition issue in HAProxy [SPOE config](https://github.com/corazawaf/coraza-spoa/blob/main/docker/haproxy/coraza.cfg).

Reply:
Returning a `bool` seems fine - from what it looks like some fields are required and some aren't so the caller needs to determine whether it's an error or not, not this utility. Or otherwise we would need two versions of each, `getFoo` and `getFooRequired` or similar, but maybe too much

Reply:
Still we might need to learn about the error on retrieval. How about

```go
var ErrNotFound = errors.New("key not found")

func getString(args map[string]interface{}, key string) (string, error) {
	argVal, exist := args[key]
	if !exist {
		return "", errNotFound
	}
	if argVal == nil {
		return "", nil
	}
	val, ok := argVal.(string)
	if !ok {
		return "", fmt.Errorf("invalid argument for %s, string expected, got %v", key, argVal)
	}
	return val, nil
}
```

so that we can flag when the key isn't in the map and return a `ErrNotFound` to the upstream function letting them decide what happens if not found and finally another error in case the data isn't correct (hence programatic error).

Reply:
> the caller needs to determine whether it's an error or not
Already does:
```
	req.method, err = getString(args, "method")
	if err != nil {
		return nil, err
	}
	req.path, err = getString(args, "path")
	if err != nil {
		fmt.Println(err.Error())
	}
	req.body, _ = getByteArray(args, "body")
```